### ISSUE TITLE
HTTP server with authentication

### ISSUE THREAD

**Issue by Eki474** [jtpereyda/boofuzz#477]
Hi,

I'm searching to fuzz my HTTP server and its API, this server is protected by an authentication phase. Authentication process is by HTTP request containing username and password. Server response will contain an authenticated token the client will need for all subsequent requests (integrated in "Authorization" header field).

In order to fuzz the API, Boofuzz has to authenticate itself to the server. How can I do that ?

**Comment by Eki474** [comment]
I started a solution using what is described in #306 but I am not sure it will give the results I want.

**Comment by SR4ven** [comment]
Would it be possible to handle the authentication before the fuzzing session starts? That way you could set the token statically at the correct field.

Another way could be reauthorising for every test case in a pre_send callback.

**Comment by jtpereyda** [comment]
You would be able to send an initial auth exchange, although it will add some back and forth overhead and authenticating right before fuzzing (perhaps with a simple script that runs at the beginning of the fuzz script) may be more efficient.

**Comment by Eki474** [comment]
Hi,

@SR4ven thanks a lot for your message, I used my findings and your answer and came to a solution:

1. Authentication process before starting to fuzz with token retrieval and storage
2. Each Boofuzz message definition contains a static named field to be used for the token (`TOKEN_VARIABLE_NAME`)
3. Session `pre_send` callback adds token to every message
4. Session `post_test_case` callback is used to renew the token between each test

_Note: my token is valid for 1 hour. After one hour, the token expires and the session is closed. Which is why I need to authenticate myself again in between testcases using `post_test_case` callback._

As an example, here is a very basic implementation for the callbacks (using `requests` module for HTTP requests).

```python
def session_pre_send_callback(target, fuzz_data_logger, session, sock):
    session.fuzz_node.names[TOKEN_VARIABLE_NAME]._value = f"Authorization: Bearer {auth_token}\r\n"
    return session.fuzz_node.render()
```

```python
def session_post_test_case_callback_admin(target, fuzz_data_logger, session, sock):
    auth_response = requests.post("http://server_address/login", data={
        "_username": "my_username",
        "_password": "my_password"
    })

    global auth_token
    auth_token = auth_response.json()['token']

```

This solves my issue. Thanks again for your help.

**Comment by jtpereyda** [comment]
@Eki474 Thank you for sharing your solution!

**Comment by User-debbie** [comment]
@llabadie I am facing the same situation. Thanks for your guidance.  But when I tried it I've encountered a problem. 
`[2021-11-22 17:20:24,424]     Error!!!! A custom pre_send callback function raised an uncought error.
                              Traceback (most recent call last):
                                File "C:\Users\l00610226\AppData\Local\Programs\Python\Python37\lib\site-packages\boofuzz\monitors\callback_monitor.py", line 44, in pre_send
                                  f(target=target, fuzz_data_logger=fuzz_data_logger, session=session, sock=target)
                                File "D:/test/working-example4.py", line 74, in give_auth_code
                                  print("old_authorization_code_value:"+str(session.fuzz_node.names['Request.Authorization.authorization_code']._value))
                              **AttributeError: 'Static' object has no attribute '_value'**
                              '`
It says *AttributeError: 'Static' object has no attribute '_value. Where can I find fuzz_node definition? and how can I give this "auth_code" to every request before they are sent?

**Comment by cq674350529** [comment]
@User-debbie Maybe you can refer to this: https://github.com/jtpereyda/boofuzz/issues/367 . Since the boofuzz has evolved, some methods/fields(e.g. `_value` field) have been changed.

**Comment by User-debbie** [comment]
@cq674350529 Thanks~I'll read them first then try it

**Comment by User-debbie** [comment]
@cq674350529 Thanks~I have solved the problem by populating field `_default value`, the whole sentence is `session.fuzz_node.names['Request.Authorization.authorization_code']._default_value = auth_token`, then variable `auth_token` is given to this node in `pre_send_callbacks`. Now I'm trying to figure out how to use variable mutation strategies, as I've noticed that primitive `s_string` would fill the node with value of a fixed database, called `bad numbers`? I would like it to be REAL FUZZ, and REAL RAMDOM, but I haven't found the solution. Maybe I should use other primitives? e.g. `s_random`?

**Comment by User-debbie** [comment]
or maybe it is already REAL RANDOM when I use primitive `s_string`?I am fuzzing my API successfully for the first time, and I don't know how many test case it will generate, the console says: **Total: xxxx of many**
![image](https://user-images.githubusercontent.com/37677388/142990273-2621f49c-58a3-4c45-8036-ec0faf6b36f0.png)

 @cq674350529

**Comment by cq674350529** [comment]
> @cq674350529 Now I'm trying to figure out how to use variable mutation strategies, as I've noticed that primitive `s_string` would fill the node with value of a fixed database, called `bad numbers`? I would like it to be REAL FUZZ, and REAL RANDOM, but I haven't found the solution. Maybe I should use other primitives? e.g. `s_random`?

If the type of `variable` to mutate is `string`, then use `s_string()` is ok. `s_string()` has some inbuilt strategy, and the generated mutate data is based on `_fuzz_library`, `long_string_seeds`, and length variation.

By the way, if you have questions, you can open a new issue.

**Comment by SR4ven** [comment]
@User-debbie for random data you can use `s_random` or `RandomData`.

About "xxxx of many" test cases, that's because of the combinatorial fuzzing. It permutes over over all primitives so number of test cases rises really quick and is hard to calculate. It's more or less generating endless amounts of test cases.
If you don't want this feature, you can disable it by `setting max_depth=1` in [Session.fuzz()](https://boofuzz.readthedocs.io/en/stable/source/Session.html#boofuzz.Session.fuzz).

Feel free to open new issues or ask questions on [Gitter](https://gitter.im/jtpereyda/boofuzz).

Thanks for your help @cq674350529!

**Comment by User-debbie** [comment]
Thanks @cq674350529 @SR4ven for your help! The `variable` I am fuzzing currently is `string`, and I'll try `setting max_depth=1` in order to control the test cases it generates.
